shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java

```diff
@@ -338,7 +358,279 @@ public void run() {
   }
 
   @Keep
-  final class ImageTextureRegistryEntry implements TextureRegistry.ImageTextureEntry {
+  final class ImageRawSurfaceTexture
+      implements TextureRegistry.RawSurfaceTextureEntry, TextureRegistry.ImageProducerTextureEntry {
+    private static final String TAG = "ImageRawSurfaceTexture";
+    private static final int MAX_IMAGES = 4;
+
+    private final long id;
+
+    private boolean released;
+    private boolean ignoringFence = false;
+
+    private int requestedWidth = 0;
+    private int requestedHeight = 0;
+
+    private ImageReader activeReader;
+    private ImageReader toBeClosedReader;
+    private Image latestImage;
+
+    private final Handler onImageAvailableHandler = new Handler();
+    private final ImageReader.OnImageAvailableListener onImageAvailableListener =
+        new ImageReader.OnImageAvailableListener() {
+          @Override
+          public void onImageAvailable(ImageReader reader) {
+            Image image = null;
+            try {
+              image = reader.acquireLatestImage();
+            } catch (IllegalStateException e) {
+              Log.e(TAG, "onImageAvailable acquireLatestImage failed: " + e.toString());
+            }
+            if (image == null) {
+              return;
+            }
+            onImage(image);
+            maybeCloseReader();
```

Comment:
nit: Maybe this should go in a finally block? I'm not sure of the API - maybe IllegalStateException is the only possible outcome and it's safe.